### PR TITLE
refactor: [ANDROSDK-1712] replace rxjava login call

### PR DESCRIPTION
**Pull request by dan-dhis2**
fix:
- LogInCallUnitShould.kt to handle the refactor
- CoroutineAPICallExecutorMock.kt now it handles D2Error gently

refactor: to coroutines
- UserModuleImpl.kt
- UserCall.kt
- LogInCall.kt
- some isolated calls to the new coroutines functions
- UserService.kt to remove retrofit's "Call"
- OpenIDConnectHandlerImpl.kt
- rxjava from LogInCallUnitShould.kt
- UserCallShould.kt

### CODE REVIEW

File: core/src/main/java/org/hisp/dhis/android/core/user/internal/LogInCall.kt [hunk 1]
```diff
@@ -84,15 +76,15 @@ internal class LogInCall @Inject internal constructor(
...
-            UserIdAuthenticatorHelper.basic(username!!, password!!),
-            UserFields.allFieldsWithoutOrgUnit(null),
-        )
-
-        val credentials = Credentials(username, trimmedServerUrl!!, password, null)
+        val credentials = Credentials(username!!, trimmedServerUrl!!, password, null)
 
         return try {
-            val user = apiCallExecutor.executeObjectCallWithErrorCatcher(authenticateCall, apiCallErrorCatcher)
+            val user = coroutineAPICallExecutor.wrap {
```
Comment (vgarciabnz) [inline core/src/main/java/org/hisp/dhis/android/core/user/internal/LogInCall.kt:82]: The `apiCallErrorCatcher` is removed. It must be passed as a parameter to the `wrap` function.

File: core/src/main/java/org/hisp/dhis/android/core/user/internal/LogInCall.kt [hunk 2]
```diff
@@ -182,7 +172,7 @@ internal class LogInCall @Inject internal constructor(
 
         var credentials: Credentials? = null
         return try {
-            val user = apiCallExecutor.executeObjectCallWithErrorCatcher(authenticateCall, apiCallErrorCatcher)
+            val user = coroutineAPICallExecutor.wrap { authenticateCall }.getOrThrow()
```
Comment (vgarciabnz) [inline core/src/main/java/org/hisp/dhis/android/core/user/internal/LogInCall.kt:175]: The `authenticateCall`is executed at the time it is declared. Before, it was a `Callable` so it was executed when its method `call` was invoked; but now it is a suspend function.
The actual invokation must be wrapped by the `wrap` function.

File: gradle/wrapper/gradle-wrapper.properties [hunk 3]
```diff
@@ -1,5 +1,6 @@
+#Thu Aug 31 00:09:15 CEST 2023
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-8.2.1-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.0-bin.zip
```
Comment (vgarciabnz) [inline gradle/wrapper/gradle-wrapper.properties:4]: Why this downgrade?

**Review by vgarciabnz — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/hisp/dhis/android/core/user/internal/LogInCall.kt [hunk 4]
```diff
@@ -169,20 +160,20 @@ internal class LogInCall @Inject internal constructor(
...
 
-        val authenticateCall = userService.authenticate(
-            "Bearer ${openIDConnectState.idToken}",
-            UserFields.allFieldsWithoutOrgUnit(versionManager.getVersion()),
-        )
-
         var credentials: Credentials? = null
         return try {
-            val user = apiCallExecutor.executeObjectCallWithErrorCatcher(authenticateCall, apiCallErrorCatcher)
+            val user = coroutineAPICallExecutor.wrap {
```
Comment (vgarciabnz) [inline core/src/main/java/org/hisp/dhis/android/core/user/internal/LogInCall.kt:171]: The `apiCallErrorCatcher` is missing

**Review by vgarciabnz — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/hisp/dhis/android/core/user/internal/LogInCall.kt (hunk 4 above)
Comment (dan-dhis2) [reply core/src/main/java/org/hisp/dhis/android/core/user/internal/LogInCall.kt]: Done!

**Review by vgarciabnz — APPROVED** [review]

**Pull request merged**